csharp/src/Ice/IceDiscovery/Locator.cs

```diff
@@ -154,43 +200,52 @@ internal class LocatorRegistry : ILocatorRegistry
                     return null;
                 }
 
-                foreach ((string key, HashSet<string> ids) in _replicaGroups)
+                foreach (((string ReplicaGroupId, Protocol Protocol) key, HashSet<string> adapterIds)
+                    in _replicaGroups)
                 {
-                    try
+                    if (key.Protocol == protocol)
                     {
-                        // We retrieve and clone this proxy _only_ for its protocol and encoding. All the other
-                        // information in the proxy is wiped out or replaced.
-
-                        IObjectPrx proxy = _adapters[ids.First()];
-                        proxy = proxy.Clone(IObjectPrx.Factory,
-                                            endpoints: ImmutableArray<Endpoint>.Empty,
-                                            identity: identity,
-                                            location: ImmutableArray.Create(key));
-                        proxy.IcePing();
-                        return proxy;
-                    }
-                    catch
-                    {
-                        // Ignore.
+                        // We retrieve this proxy only for its protocol - everything else gets replaced by the Clone
+                        // below.
+                        IObjectPrx dummy = _adapters[(adapterIds.First(), protocol)];
```

Comment:
Couldn't the adapter ID not be in the dictionary? This is also related to the "TODO: inconsistency" above... 

We should decide if the IceDiscovery locator and locator registry interfaces should be internal only and not accessible from the outside world. In this case there should be no inconsistency and we could assert if the adapter isn't found. If accessible to the outside world, there could be inconsistencies and we should also deal with it here. We could also change the _adapters dictionary to keep track of the registered replica group ID and remove the adapter from the replica group or raise an exception if the adapter is unregistered with inconsistent replica group ID.

Reply:
I think Locator and LocatorRegistry should be only colocated and not accessible remotely. I'll fix the TODO inconsistency and add a comment.